### PR TITLE
feat: My study migration

### PR DESCRIPTION
**Pull request by maghirardelli**
Issue #, if available: GALI-1352

Description of changes: added new migration service and controller that has two APIs: 
1. **PUT /migrate/mystudies** - accepts a request body of studyId and uid mappings to migrate those StudyIds to be owned by those uids. Uids must be for non internal users. StudyIds must not already be owned by uids. 
2. **GET /migrate/mystudies** - lists all studies of the category My Studies. Both APIs can only be called by admin users. 
 
Unit tests have been written for all new/modified methods. Integ tests will come later after Sanket's native Cognito PR.

Checklist:

- [x] Have you successfully deployed to an AWS account with your changes?
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you successfully tested with your changes locally?
- [x] Have you updated openapi.yaml if you made updates to API definition (including add, delete or update parameter and request data schema)?

AS review ticket id:

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    const studyOperationService = await this.service('studyOperationService');
+    const userService = await this.service('userService');
+    const studyService = await this.service('studyService');
+
+    const migrationResults = [];
+    await Promise.all(
+      migrationMappings.map(async item => {
+        const endUser = item.uid;
+        const studyId = item.studyId;
+        // console.log(result1);
```
Comment (SanketD92) [inline addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js]: redundant comment

File: addons/addon-base-raas/packages/base-raas-rest-api/lib/controllers/migration-controller.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+  // ===============================================================
+  //  GET /mystudies (mounted to /api/migrate)
+  // ===============================================================
+  router.get(
+    '/mystudies',
+    wrap(async (req, res) => {
+      const migrationService = await context.service('migrationService');
+      const requestContext = res.locals.requestContext;
+      const result = await migrationService.listMyStudies(requestContext);
```
Comment (SanketD92) [inline addons/addon-base-raas/packages/base-raas-rest-api/lib/controllers/migration-controller.js:42]: Does the response make it evident which studies require migration, and which internal user the resource belonged to? These values in the response body will help a lot for admins to figure out what to migrate:

```
studyId: <sampleStudyId>,
uid: <sampleUidValue>,
username: <sampleUsername>,
authProvider: <respectiveAuthProviderId / 'internal'> 

```

File: addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    this._query = () => dbService.helper.query().table(studiesTable);
+    this._deleter = () => dbService.helper.deleter().table(studiesTable);
+    this._scanner = () => dbService.helper.scanner().table(studiesTable);
+
+    this.categoryIndex = this.settings.get(settingKeys.categoryIndexName);
+  }
+
+  async migrateMyStudiesPermissions(requestContext, migrationMappings) {
+    if (!isAdmin(requestContext)) {
+      throw this.boom.forbidden("You don't have permission to migrate My Studies", true);
```
Comment (nguyen102) [inline addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js]: Let's be more specific and say "You need admin permissions to migrate My Studies"

File: addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    const studyService = await this.service('studyService');
+
+    const migrationResults = [];
+    await Promise.all(
+      migrationMappings.map(async item => {
+        const endUser = item.uid;
+        const studyId = item.studyId;
+        // console.log(result1);
+        if (await userService.isInternalAuthUser(endUser)) {
+          throw this.boom.forbidden('You cannot migrate My Studies to internal user', true);
```
Comment (nguyen102) [inline addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js]: These errors should be a badRequest, not a permission error.
`this.boom.badRequest`

File: addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        }
+
+        // check that end user is not already owner
+        if (await studyService.isStudyAdmin(requestContext, studyId, endUser)) {
+          throw this.boom.forbidden('You cannot migrate My Studies to current owner', true);
```
Comment (nguyen102) [inline addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js]: These errors should be a badRequest, not a permission error.
this.boom.badRequest.

`Current study is already owned by ${endUser}`

File: addons/addon-base-raas/packages/base-raas-services/lib/migration/__tests__/migration-service.test.js [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+      // BUILD
+      const requestContext = createAdminContext();
+      const migrationMapping = [
+        { studyId: 'testStudy1', uid: 'testUser1' },
+        { studyId: 'testStudy2', uid: 'testUser2' },
+      ];
+
+      // OPERATE
+      const results = await service.migrateMyStudiesPermissions(requestContext, migrationMapping);
+      //   console.log(results);
```
Comment (nguyen102) [inline addons/addon-base-raas/packages/base-raas-services/lib/migration/__tests__/migration-service.test.js]: delete

File: addons/addon-base-raas/packages/base-raas-services/lib/migration/__tests__/migration-service.test.js [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    });
+
+    it('should pass correct updateRequest', async () => {
+      // BUILD
+      const requestContext = createAdminContext();
+      const migrationMapping = [{ studyId: 'testStudy1', uid: 'testUser1' }];
+
+      // OPERATE
+      await service.migrateMyStudiesPermissions(requestContext, migrationMapping);
+      //   console.log(results);
```
Comment (nguyen102) [inline addons/addon-base-raas/packages/base-raas-services/lib/migration/__tests__/migration-service.test.js]: delete

File: addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        }
+
+        const result = await studyOperationService.updatePermissions(requestContext, studyId, {
+          usersToAdd: [{ uid: endUser, permissionLevel: 'admin' }],
+          usersToRemove: [{ uid: '*', permissionLevel: 'admin' }],
```
Comment (nguyen102) [inline addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js:68]: I'm not familiar with study permissions. What does a `uuid` of `*` mean? In what cases does a study have an owner of `*`?

Comment (maghirardelli) [reply addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js:68]: This is a new convention. The `*` means to remove all users like a wildcard. In the case of the My Studies, this will only be one user. But we don't necessarily know the current user so we need to put something. The logic in the methods file later adds whoever was the old admin of the My Study to be the value in place of `*`. 

TLDR: `*` is never the owner--just a wildcard placeholder.

File: addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js (hunk 3 above)
Comment (maghirardelli) [reply addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js]: Good idea! Adopting this wording.

File: addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js (hunk 5 above)
Comment (maghirardelli) [reply addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js]: Got it! Thank you!

**Review by nguyen102 — DISMISSED** [review]

File: openapi.yaml [hunk 9]
```diff
@@ -2147,6 +2147,34 @@ paths:
       responses:
         '200':
           description: 'Return success update message'
+  /api/migrate/mystudies:
```
Comment (rsmayda) [inline openapi.yaml]: nit: should apis be in kebab-case (my-studies)?

File: addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        const endUser = item.uid;
+        const studyId = item.studyId;
+        if (await userService.isInternalAuthUser(endUser)) {
+          throw this.boom.badRequest('You cannot migrate My Studies to internal user', true);
+        }
+
+        // check that end user is not already owner
+        if (await studyService.isStudyAdmin(requestContext, studyId, endUser)) {
+          throw this.boom.badRequest(`Current study is already owned by ${endUser}`, true);
+        }
```
Comment (rsmayda) [inline addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js]: should this be a failure? effectively this request is to change the study ownership to `endUser` but this check is saying that `endUser` is already the owner. Please correct me if wrong!

File: addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+      throw this.boom.forbidden('You need admin permissions to list all My Studies in this environment', true);
+    }
+
+    const studyService = await this.service('studyService');
+    const userService = await this.service('userService');
+
+    const result = await this._query()
+      .index(this.categoryIndex)
+      .key('category', 'My Studies')
+      .limit(1000)
```
Comment (rsmayda) [inline addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js:87]: 1. Will this API be hit from the UI? 
2. How likely it is to hit the limit?

File: addons/addon-base-raas/packages/base-raas-services/lib/migration/__tests__/migration-service.test.js [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+
+      // CHECK
+      expect(studyOperationService.updatePermissions).toHaveBeenCalledWith(requestContext, 'testStudy1', {
+        usersToAdd: [{ uid: 'testUser1', permissionLevel: 'admin' }],
+        usersToRemove: [{ uid: '*', permissionLevel: 'admin' }],
+      });
+    });
+  });
+
+  describe('listMyStudies', () => {
```
Comment (rsmayda) [inline addons/addon-base-raas/packages/base-raas-services/lib/migration/__tests__/migration-service.test.js:156]: Not testing happy path at all

File: addons/addon-base-raas/packages/base-raas-services/lib/study/helpers/entities/study-permissions-methods.js [hunk 13]
```diff
@@ -52,15 +52,23 @@ function isAdmin(studyPermissionsEntity, uid) {
 function applyUpdateRequest(studyPermissionsEntity, updateRequest) {
   const entity = studyPermissionsEntity;
 
+  if (updateRequest.usersToRemove[0].uid === '*') {
+    // For mitigation of old internal users My Studies
+    const level = `${updateRequest.usersToRemove[0].permissionLevel}Users`;
+    const oldUser = entity[level][0];
+    updateRequest.usersToRemove = [{ uid: oldUser, permissionLevel: 'admin' }];
```
Comment (rsmayda) [inline addons/addon-base-raas/packages/base-raas-services/lib/study/helpers/entities/study-permissions-methods.js:59]: What if there are multiple "old" admins?

File: addons/addon-base-raas/packages/base-raas-services/lib/study/helpers/entities/__tests__/study-permissions-methods.test.js [hunk 14]
```diff
@@ -55,5 +55,54 @@ describe('study permissions methods', () => {
         writeonlyUsers: ['u-4'],
       });
     });
+
+    it('should apply requested changes for migration', () => {
```
Comment (rsmayda) [inline addons/addon-base-raas/packages/base-raas-services/lib/study/helpers/entities/__tests__/study-permissions-methods.test.js:59]: could you add test for multiple adminUsers

File: addons/addon-base-raas/packages/base-raas-services/lib/study/study-service.js [hunk 15]
```diff
@@ -388,7 +388,8 @@ class StudyService extends Service {
       throw this.boom.forbidden('Permissions cannot be set for studies in the "Open Data" category', true);
     }
 
-    if (isMyStudies(studyEntity)) {
+    // If study entity is My Study and not part of a migration request, throw error
+    if (isMyStudies(studyEntity) && updateRequest.usersToRemove[0].uid !== '*') {
```
Comment (rsmayda) [inline addons/addon-base-raas/packages/base-raas-services/lib/study/study-service.js]: when updating permission we can now use `*`  for any update permissions call; is that the intention?

Also we are saying MyStudies MUST have  `*` in "toRemove" is that the intention?

File: addons/addon-base-raas/packages/base-raas-services/lib/study/__tests__/study-service.test.js [hunk 16]
```diff
@@ -1294,4 +1294,103 @@ describe('studyService', () => {
       expect(retVal).toEqual(false);
     });
   });
+
+  describe('updatePermissions', () => {
+    it('should throw error if My Study but not migration request', async () => {
```
Comment (rsmayda) [inline addons/addon-base-raas/packages/base-raas-services/lib/study/__tests__/study-service.test.js:1299]: that isn't so clear it is probably clearer to say: 
```suggestion
    it('should throw error if My Study and usersToRemove is not *', async () => {
```

File: addons/addon-base-raas/packages/base-raas-services/lib/study/__tests__/study-service.test.js [hunk 17]
```diff
@@ -1294,4 +1294,103 @@ describe('studyService', () => {
       expect(retVal).toEqual(false);
     });
   });
+
+  describe('updatePermissions', () => {
```
Comment (rsmayda) [inline addons/addon-base-raas/packages/base-raas-services/lib/study/__tests__/study-service.test.js:1298]: Could you add another use-case for using the `*` outside of the MyStudy context

File: openapi.yaml (hunk 9 above)
Comment (maghirardelli) [reply openapi.yaml]: Will change.

File: addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js (hunk 10 above)
Comment (maghirardelli) [reply addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js]: The migration is failing was my logic. The request is failing because what was requested was already up to date. Would you prefer a different boom error?

File: addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js (hunk 11 above)
Comment (maghirardelli) [reply addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js:87]: 1. The API will not be hit from the UI.
2. This is the same limit on the method called from GET /studies (`list()` in `study-service.js`). If we are accepting this limit there, I thought it would be okay to accept the limit here.

File: addons/addon-base-raas/packages/base-raas-services/lib/study/helpers/entities/study-permissions-methods.js (hunk 13 above)
Comment (maghirardelli) [reply addons/addon-base-raas/packages/base-raas-services/lib/study/helpers/entities/study-permissions-methods.js:59]: My Studies by design only have one admin. A user cannot share them and before now, could not transition ownership. Therefore, the AdminUsers list will only have one user ever and none of the other permissions lists will have any users as well.

File: addons/addon-base-raas/packages/base-raas-services/lib/study/helpers/entities/__tests__/study-permissions-methods.test.js (hunk 14 above)
Comment (maghirardelli) [reply addons/addon-base-raas/packages/base-raas-services/lib/study/helpers/entities/__tests__/study-permissions-methods.test.js:59]: See above comment, but there will never be multiple adminUsers as My Studies by design are only owned and can only be accessed by one user (formerly the one who created it; now, the one who created it or whoever it is migrated to).

File: addons/addon-base-raas/packages/base-raas-services/lib/study/study-service.js (hunk 15 above)
Comment (maghirardelli) [reply addons/addon-base-raas/packages/base-raas-services/lib/study/study-service.js]: By design, we have never before allowed changing permissions on My Study because they are not sharable. So, yes, MyStudies MUST have `*` in "toRemove" because that is what the API controller passes in. It is the only API that does this.
Theoretically, a user could put `*` in `userToRemove` in an update permissions API call body which might cause unexpected behavior. We can either assume that won't happen because `*` has never been a valid input before or we can build some sort of check to make sure that the only API that can pass `*` to update permissions in `userToRemove` is the my studies migration API. What is acceptable/required in your opinion?

File: addons/addon-base-raas/packages/base-raas-services/lib/study/__tests__/study-service.test.js (hunk 16 above)
Comment (maghirardelli) [reply addons/addon-base-raas/packages/base-raas-services/lib/study/__tests__/study-service.test.js:1299]: I am going to combine both messages, as they *should* be synonymous with each other because the only time `usersToRemove` should be `*` is when it is a migration request.

File: addons/addon-base-raas/packages/base-raas-services/lib/study/__tests__/study-service.test.js (hunk 17 above)
Comment (maghirardelli) [reply addons/addon-base-raas/packages/base-raas-services/lib/study/__tests__/study-service.test.js:1298]: There should not be a use-case for using `*` outside of the MyStudy context.

File: addons/addon-base-raas/packages/base-raas-services/lib/study/study-service.js (hunk 15 above)
Comment (rsmayda) [reply addons/addon-base-raas/packages/base-raas-services/lib/study/study-service.js]: I would put in logic to prevent the use of `*` unless we explicitly want that functionality

File: addons/addon-base-raas/packages/base-raas-services/lib/study/__tests__/study-service.test.js (hunk 17 above)
Comment (rsmayda) [reply addons/addon-base-raas/packages/base-raas-services/lib/study/__tests__/study-service.test.js:1298]: There is nothing from the code preventing this use-case from happening so it is a side-effect we need to test

File: addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js (hunk 10 above)
Comment (rsmayda) [reply addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js]: since the request is both `endUser` and `studyId` for a user to action this error they would need both, right?

Comment (maghirardelli) [reply addons/addon-base-raas/packages/base-raas-services/lib/migration/migration-service.js]: Good catch. I'll include both.

File: addons/addon-base-raas/packages/base-raas-services/lib/study/study-service.js (hunk 15 above)
Comment (maghirardelli) [reply addons/addon-base-raas/packages/base-raas-services/lib/study/study-service.js]: Got it! Will do.

File: addons/addon-base-raas/packages/base-raas-services/lib/migration/__tests__/migration-service.test.js (hunk 12 above)
Comment (maghirardelli) [reply addons/addon-base-raas/packages/base-raas-services/lib/migration/__tests__/migration-service.test.js:156]: Added this test

File: addons/addon-base-raas/packages/base-raas-services/lib/study/__tests__/study-service.test.js (hunk 17 above)
Comment (maghirardelli) [reply addons/addon-base-raas/packages/base-raas-services/lib/study/__tests__/study-service.test.js:1298]: Removed the part that relies on `*` as migration indicator and disallowed `*` to be part of the update request so there is no need to test anymore

File: addons/addon-base-raas/packages/base-raas-services/lib/study/helpers/entities/study-permissions-methods.js [hunk 18]
```diff
@@ -71,6 +71,13 @@ function applyUpdateRequest(studyPermissionsEntity, updateRequest) {
   });
 }
 
+function isValidUpdateRequest(updateRequest) {
+  // Check to make sure the wildcard is not used as a uid in usersToRemove
+  const wildcardEntries = updateRequest.usersToRemove.filter(entry => entry.uid === '*');
```
Comment (rsmayda) [inline addons/addon-base-raas/packages/base-raas-services/lib/study/helpers/entities/study-permissions-methods.js]: I think using `any` would make this simpler

```suggestion
  return updateRequest.usersToRemove.any(entry => entry.uid === '*');
```

File: addons/addon-base-raas/packages/base-raas-services/lib/study/study-permission-service.js [hunk 19]
```diff
@@ -337,6 +338,14 @@ class StudyPermissionService extends Service {
         { studyEntity, studyPermissionsEntity },
       );
 
+      // Ensure only migration requests can have wildcards in the updateRequest
+      if (!('isMigration' in requestContext) && !isValidUpdateRequest(updateRequest)) {
```
Comment (rsmayda) [inline addons/addon-base-raas/packages/base-raas-services/lib/study/study-permission-service.js]: I think this is a bit clearer
```suggestion
      if (!requestContext.isMigration && !isValidUpdateRequest(updateRequest)) {
```

File: addons/addon-base-raas/packages/base-raas-services/lib/study/study-service.js [hunk 20]
```diff
@@ -389,7 +389,7 @@ class StudyService extends Service {
     }
 
     // If study entity is My Study and not part of a migration request, throw error
-    if (isMyStudies(studyEntity) && updateRequest.usersToRemove[0].uid !== '*') {
+    if (isMyStudies(studyEntity) && !('isMigration' in requestContext)) {
```
Comment (rsmayda) [inline addons/addon-base-raas/packages/base-raas-services/lib/study/study-service.js]:
```suggestion
    if (isMyStudies(studyEntity) && !requestContext.isMigration) {
```

File: addons/addon-base-raas/packages/base-raas-services/lib/study/helpers/entities/study-permissions-methods.js (hunk 18 above)
Comment (maghirardelli) [reply addons/addon-base-raas/packages/base-raas-services/lib/study/helpers/entities/study-permissions-methods.js]: I think Javascript uses `some` for this purpose. It did make it simpler! Thanks for pointing out there may be a more efficient array method.

**Review by rsmayda — APPROVED** [review]
